describe cursor_pos ambiguity and bump protocol to 5.2

minrk opened the pull request:
cursor_pos with characters after 0x10000 is officially ambiguous with respect to surrogate pairs in versions 5.1 and earlier due to a widespread bug in javascript frontends.

Thus bumps the protocol to 5.2 with the sole change that cursor_pos is no longer ambiguous (i.e. clients that correctly implement cursor_pos in 5.1 can bump to 5.2 with no other changes).

cf #259

cross-refs for frontend implementations:

- https://github.com/sagemathinc/cocalc/issues/1994
- https://github.com/jupyterlab/jupyterlab/issues/2255
- https://github.com/nteract/nteract/issues/1706
- https://github.com/jupyter/notebook/issues/2509

cc @stevengj @rgbkrk @Carreau @blink1073 @williamstein

Carreau commented on docs/messaging.rst:
```diff
@@ -1370,12 +1385,44 @@ handlers should set the parent header and publish status busy / idle,
+Many frontends, especially those implemented in javascript,
+reported cursor_pos as the interpreter's string index,
+which is not the same as the unicode character offset if the interpreter uses UTF-16 (e.g. javascript or Python 2 on macOS),
+which stores astral-plane characters such as ``𝐚`` as surrogate pairs,
```
precise  `U+1D41A` ? Because I'm going to assume somewhere it's not going to render correctly ?

minrk replied:
Good idea

Carreau commented on docs/messaging.rst:
```diff
@@ -1370,12 +1385,44 @@ handlers should set the parent header and publish status busy / idle,
+- JupyterLab
+- nteract
+- CoCalc
+- Jupyter Console and QtConsole with Python 2 on macOS
```
Should we precise versions ?

stevengj replied:
probably on Windows, too

minrk replied:
Confirmed on Windows Python 2 as well.

stevengj commented on docs/messaging.rst:
```diff
@@ -547,6 +547,14 @@ Message type: ``inspect_request``::
+.. versionchanged:: 5.2
+
+    Due to a widespread bug in many frontends, ``cursor_pos``
+    in versions prior to 5.2 is ambiguous in the presence of astral-plane.
```
missing a noun:
```
in the presence of "astral-plane" characters
```
(added scare quotes around "astral-plane" since this is a slang term)

stevengj replied:
similar below

stevengj replied:
maybe also link the github issue for the technical details?

stevengj commented:
In `kernel_info_reply` the kernel sends what version of the protocol it implements.  Is the highest supported protocol version sent in the reverse direction (from frontend to kernel) anywhere?

minrk commented:
> Is the highest supported protocol version sent in the reverse direction (from frontend to kernel) anywhere?

This ought to be the 'version' field in the request header, I believe.

blink1073 commented:
Currently `5.0` in [JupyterLab](https://github.com/jupyterlab/jupyterlab/blob/aa825677987983d798920b42879d7657d7f7ccd3/packages/services/src/kernel/messages.ts#L30) and [Notebook 5.0](https://github.com/jupyter/notebook/blob/cd47dac0ff5ddae4261eea4540ec5d0194984439/notebook/static/services/kernels/kernel.js#L83).

stevengj commented:
@minrk, currently `kernel_info_request` is documented as `content = {}`.  Adding a version field would be a good addition.  Can that be added to protocol version 5.2 as well?

minrk commented:
Came back to this. I believe I've addressed comments now, and issued PRs against jupyter/notebook and jupyterlab to handle the transform and bump the version to 5.2.

stevengj commented on docs/messaging.rst:
```diff
@@ -1370,12 +1385,48 @@ handlers should set the parent header and publish status busy / idle,
+taking up two indices instead of one, causing a unicode offset
+drift of one per astral-plane character.
+Not all frontends have this behavior, however,
+and after JSON serialization information about encoding was used
```
`about which encoding was used`

minrk replied:
Good catch, thanks!

stevengj commented:
cross-ref: nteract/hydrogen#807
